homeassistant/components/zamg/config_flow.py

```diff
@@ -0,0 +1,136 @@
+                        CONF_STATION_ID, default=int(closest_station_id)
+                    ): vol.In(
+                        {
+                            int(station): f"{stations[station][2]} ({station})"
+                            for station in stations
+                        }
+                    )
+                }
+            )
+            return self.async_show_form(step_id="user", data_schema=schema)
+
+        station_id = str(user_input[CONF_STATION_ID])
+
+        # Check if already configured
+        await self.async_set_unique_id(station_id)
+        self._abort_if_unique_id_configured()
+
+        try:
+            self._client.set_default_station(station_id)
+            await self._client.update()
+        except (ValueError, TypeError) as err:
+            LOGGER.error("Config_flow: Received error from ZAMG: %s", err)
+            errors["base"] = "cannot_connect"
+            return self.async_abort(
+                reason="cannot_connect", description_placeholders=errors
+            )
+
+        return self.async_create_entry(
+            title=user_input.get(CONF_NAME) or self._client.get_station_name,
+            data={CONF_STATION_ID: station_id},
+        )
+
+    async def async_step_import(self, config: dict[str, Any]) -> FlowResult:
+        """Handle ZAMG configuration import."""
+        station_id = str(config.get(CONF_STATION_ID))
+        station_name = config.get(CONF_NAME)
+        # create issue every time after restart
+        # parameter is_persistent seems not working
+        async_create_issue(
+            self.hass,
+            DOMAIN,
+            "deprecated_yaml",
+            breaks_in_ha_version="2023.1.0",
+            is_fixable=False,
+            severity=IssueSeverity.WARNING,
+            translation_key="deprecated_yaml",
+        )
+
+        for entry in self.hass.config_entries.async_entries(DOMAIN):
+            if station_id in entry.data[CONF_STATION_ID]:
+                return self.async_abort(
+                    reason="already_configured",
+                )
+
+        if self._client is None:
+            self._client = ZamgData()
+            self._client.session = async_get_clientsession(self.hass)
+
+        if station_id not in await self._client.zamg_stations():
+            LOGGER.warning(
+                "Configured station_id %s could not be found at zamg, adding the nearest weather station instead",
+                station_id,
+            )
+            latitude = config.get(CONF_LATITUDE) or self.hass.config.latitude
+            longitude = config.get(CONF_LONGITUDE) or self.hass.config.longitude
+            station_id = await self._client.closest_station(latitude, longitude)
+
+        if not station_name:
+            await self._client.zamg_stations()
+            self._client.set_default_station(station_id)
+            station_name = self._client.get_station_name
+
+        for entry in self.hass.config_entries.async_entries(DOMAIN):
```

Comment:
This is already done on line 95. I suggest to change the order of operations to first validate the station id and then set the unique id.

Reply:
I did recommend to to it that way to abort before making a cloud request if the `station_id` is set in yaml. 
If yaml is configured with lat/Lon or city name we have no choice but to query.